### DISCUSSION TITLE
`request.body()` not works

### DISCUSSION THREAD

**Discussion by kunansy** [fastapi/fastapi#6492] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi.exceptions import RequestValidationError
from fastapi.requests import Request

...

@app.exception_handler(RequestValidationError)
async def validation_exception_handler(request: Request,
                                       exc: RequestValidationError):
    print(await request.body())
```

### Description

I made `POST /bla-bla-bla` with invalid data, when the error occurred I can't see the body of the request, swagger endlessly loads something

### Operating System

macOS

### Operating System Details

MacOS 12.1 Monterey

### FastAPI Version

0.70.1

### Python Version

3.9.9

**Comment by drake-mer** [comment]
This is because you use an async controller.

Apparently you can't print the body of a request inside an async controller.

Try to remove async and see what happens.

**Comment by panla** [comment]
use await request.json()

or in route

```python
from json import JSONDecodeError

from fastapi import Request
from fastapi.routing import APIRoute

from extensions import logger


class Route(APIRoute):
    """Extension Route and log request.json()"""

    def get_route_handler(self):
        original_route_handler = super().get_route_handler()

        async def log_request_detail(request: Request):

            logger.info('start request'.center(60, '*'))
            logger.info(f'{request.method} {request.url}')

            methods = ['POST', 'PUT', 'PATCH']
            content_type = request.headers.get('content-type')

            if request.method in methods and 'application/json' in content_type:
                try:
                    params = await request.json()
                    if params:
                        logger.info(params)
                except JSONDecodeError:
                    logger.error('encounter JSONDecodeError')
                except UnicodeDecodeError:
                    logger.error('encounter UnicodeDecodeError')
            logger.info('end request'.center(60, '*'))
            return await original_route_handler(request)

        return log_request_detail

router = APIRouter(route_class=Route)
```

**Comment by kunansy** [comment]
> This is because you use an async controller.
> 
> Apparently you can't print the body of a request inside an async controller.
> 
> Try to remove async and see what happens.

in this case there are:
```
INFO:     127.0.0.1:49432 - "GET / HTTP/1.1" 404 Not Found
INFO:     127.0.0.1:49432 - "GET /docs HTTP/1.1" 200 OK
INFO:     127.0.0.1:49432 - "GET /openapi.json HTTP/1.1" 200 OK
# the print 
<coroutine object Request.body at 0x1087081c0>
```

So I have to use async but can't

**Comment by kunansy** [comment]
> use await request.json()

all the same to `.json()`

**Comment by ahnaf-zamil** [comment]
That's strange, it just hangs the request when I try to `await req.body()` as well.

I even tried using `asyncio.ensure_future(req.body())` to see if it runs or not. It returns the HTTP response just fine, but then throws a `starlette.requests.ClientDisconnect` exception with `Task exception was never retreived`.

**Comment by panla** [comment]
in fastapi docs <https://fastapi.tiangolo.com/advanced/custom-request-and-route/>

**Comment by upeshphable** [comment]
Refer to issue #394

**Comment by kunansy** [comment]
@upeshphable thanks

**Comment by YuriiMotov** [comment]
That was fixed in [Starlette 0.28](https://www.starlette.io/release-notes/#0280-june-7-2023)
